### PR TITLE
Reflect shader stage for bind groups.

### PR DESCRIPTION
**Pull request by StarArawn**
This PR changes the shader reflection to automatically pick the correct shader stage/s.

### CODE REVIEW

File: crates/bevy_render/src/pipeline/pipeline_layout.rs [hunk 1]
```diff
@@ -25,10 +26,15 @@ impl PipelineLayout {
                         for shader_binding in shader_bind_group.bindings.iter() {
                             if let Some(binding) = bind_group
                                 .bindings
-                                .iter()
+                                .iter_mut()
                                 .find(|binding| binding.index == shader_binding.index)
                             {
-                                if binding != shader_binding {
+                                binding.shader_stage |= shader_binding.shader_stage;
+                                // Not sure we need to panic anymore here..
```
Comment (cart) [inline crates/bevy_render/src/pipeline/pipeline_layout.rs:33]: If we don't panic when binding indices dont match across shader stages, then im pretty sure wgpu will panic instead. eventually this should return an Error instead, which would make this recoverable.

**Comment by cart** [comment]
In general this looks like the right approach. Just left one comment.

**Comment by cart** [comment]
Testing on my machine now. Then we should be good to go!

**Comment by StarArawn** [comment]
@cart 
After much debugging I have a hack to fix all of the examples. In most of the 3D examples the shader being used is `forward.vert/frag` which implements the camera uniform the "both" the vertex and fragment stage. The pass node specifies that the resulting uniform must have both a Vertex and a Fragment shader stage. Before when we didn't reflect this wasn't an issue, but now its a bit more of an issue. We need a better mechanism of picking the descriptor perhaps by ignoring shader stage... Perhaps our hash implementation shouldn't account for shader stage at all, but I don't think that's right. I also don't think we should be retrieving the binding descriptor based off of name so I'm a little stuck on how I can "unhack" this.

`camera_bind_group_descriptor`:
https://github.com/bevyengine/bevy/blob/master/crates/bevy_render/src/render_graph/nodes/pass_node.rs#L73

**Comment by multun** [comment]
Can you rebase this on master? Format checking was enabled, and I believe the CI doesn't yet check for it at the commit you're at

**Comment by StarArawn** [comment]
@multun Seems like `cargo fmt` doesn't work in bevy on stable. I was able to format each file though using vscode/RSA. This is ready for review now.

**Comment by cart** [comment]
> Perhaps our hash implementation shouldn't account for shader stage at all, but I don't think that's right.

Yeah we need the hash to be a 1:1 map between "things wgpu cares about / would require a new wgpu descriptor to work"

> I also don't think we should be retrieving the binding descriptor based off of name so I'm a little stuck on how I can "unhack" this.

i think retrieving binding descriptor by index (without checking name) in a cross-shader context is probably even more shakey. what if the shader sets binding 0 to something that isnt the camera? names / conventions give us a set of "rules" we can rely on.

> I'm a little stuck on how I can "unhack" this.
Yeah this is a little weird. I think I'm fine with merging this as-is, but we should create a tracking issue to resolve the underlying problem. This probably won't be the last time it comes up.

**Comment by cart** [comment]
Follow up via #264

**Pull request merged**